modules/fx/arrow-fx/src/main/kotlin/arrow/fx/IO.kt

```diff
@@ -862,157 +865,41 @@ sealed class IO<out A> : IOOf<A> {
-   *
-   * @param release the allocated resource after the resulting [IO] of [use] is terminates.
-   *
-   * ```kotlin:ank:playground
-   * import arrow.fx.*
-   * import arrow.fx.typeclasses.ExitCase
-   *
-   * class File(url: String) {
-   *   fun open(): File = this
-   *   fun close(): Unit {}
-   *   fun content(): IO<String> =
-   *     IO.just("This file contains some interesting content!")
-   * }
-   *
-   * fun openFile(uri: String): IO<File> = IO { File(uri).open() }
-   * fun closeFile(file: File): IO<Unit> = IO { file.close() }
-   *
-   * fun main(args: Array<String>) {
-   *   //sampleStart
-   *   val safeComputation = openFile("data.json").bracketCase(
-   *     release = { file, exitCase ->
-   *       when (exitCase) {
-   *         is ExitCase.Completed -> { /* do something */ }
-   *         is ExitCase.Canceled -> { /* do something */ }
-   *         is ExitCase.Error -> { /* do something */ }
-   *       }
-   *       closeFile(file)
-   *     },
-   *     use = { file -> file.content() }
-   *   )
-   *   //sampleEnd
-   *   println(safeComputation.unsafeRunSync())
-   * }
-   *  ```
-   */
-  fun <B> bracketCase(release: (A, ExitCase<Throwable>) -> IOOf<Unit>, use: (A) -> IOOf<B>): IO<B> =
-    IOBracket(this, release, use)
-
-  /**
-   * Executes the given [finalizer] when the source is finished, either in success or in error, or if canceled.
-   *
-   * As best practice, prefer [bracket] for the acquisition and release of resources.
-   *
-   * @see [guaranteeCase] for the version that can discriminate between termination conditions
-   * @see [bracket] for the more general operation
-   */
-  fun guarantee(finalizer: IOOf<Unit>): IO<A> = guaranteeCase { finalizer }
-
-  /**
-   * Executes the given `finalizer` when the source is finished, either in success or in error, or if canceled, allowing
-   * for differentiating between exit conditions. That's thanks to the [ExitCase] argument of the finalizer.
-   *
-   * As best practice, it's not a good idea to release resources via `guaranteeCase` in polymorphic code.
-   * Prefer [bracketCase] for the acquisition and release of resources.
-   *
-   * @see [guarantee] for the simpler version
-   * @see [bracketCase] for the more general operation
-   *
-   */
-  fun guaranteeCase(finalizer: (ExitCase<Throwable>) -> IOOf<Unit>): IO<A> =
-    IOBracket.guaranteeCase(this, finalizer)
-
   internal data class Pure<out A>(val a: A) : IO<A>() {
     // Pure can be replaced by its value
     override fun <B> map(f: (A) -> B): IO<B> = Suspend { Pure(f(a)) }
 
-    // Pure can be replaced by its value
-    override fun <B> flatMap(f: (A) -> IOOf<B>): IO<B> = Suspend { f(a).fix() }
-
     override fun unsafeRunTimedTotal(limit: Duration): Option<A> = Some(a)
   }
 
-  internal data class RaiseError(val exception: Throwable) : IO<Nothing>() {
+  internal data class RaiseException(val exception: Throwable) : IO<Nothing>() {
```

Comment:
Is this gonna be redefined for BIO instead also?

Reply:
Currently in `0.10.3` there was only 1 faulty scenario for `IO` which was `Throwable`. Now, there is 2. `Throwable` and `E`, I decided to keep `RaiseError` for `E` and to refactor the original `Throwable` to `RaiseException` instead. I felt that better reflected the internals.

An `IO` for `<E, A>` is right-biased towards `A` and it short-circuits errors `E` and exceptions.